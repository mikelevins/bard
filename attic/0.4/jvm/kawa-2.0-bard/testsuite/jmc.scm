;; The example in gnu/bytecode/package.html
;; converted to Scheme by Yaroslav Kavenchuk

(define-alias ClassType gnu.bytecode.ClassType)
(define-alias Access gnu.bytecode.Access)
(define-alias Method gnu.bytecode.Method)
(define-alias CodeAttr gnu.bytecode.CodeAttr)
(define-alias Type gnu.bytecode.Type)
(define-alias Variable gnu.bytecode.Variable)
(define-alias ClassTypeWriter gnu.bytecode.ClassTypeWriter)
(define-alias ArrayClassLoader gnu.bytecode.ArrayClassLoader)

(define-alias System java.lang.System)
(define-alias Class java.lang.Class)
(define-alias Integer java.lang.Integer)

;; code

;; "public class HelloWorld extends java.lang.Object".
(define c :: ClassType (ClassType "HelloWorld"))
(*:setSuper c "java.lang.Object")
(*:setModifiers c Access:PUBLIC)


;; "public static int add(int, int)".
(define m :: Method
  (*:addMethod c "add" "(II)I" (logior Access:PUBLIC Access:STATIC)))
(define code :: CodeAttr (m:startCode))
(code:pushScope)
(code:emitLoad (code:getArg 0))
(code:emitLoad (code:getArg 1))
(code:emitAdd Type:intType)
(define resultVar :: Variable (code:addLocal Type:intType "result"))
(code:emitDup)
(code:emitStore resultVar)
(code:emitReturn)
(code:popScope)


;; Get a byte[] representing the class file.
;; We could write this to disk if we wanted.
(define classFile :: byte[] (*:writeToArray c))

;; Disassemble this class.
;; The output is similar to javap(1).
(ClassTypeWriter:print c System:out 0)

;; Load the generated class into this JVM.
;; gnu.bytecode provides ArrayClassLoader, or you can use your own.
(define cl :: ArrayClassLoader (ArrayClassLoader))
(cl:addClass "HelloWorld" classFile)
		
;; Actual invocation is just the usual reflection code.
(define helloWorldClass :: Class (cl:loadClass "HelloWorld" #t))

(define argTypes :: Class[]
  (Class[] java.lang.Integer:TYPE java.lang.Integer:TYPE))

(define result
 (as Integer
     (*:invoke (*:getMethod helloWorldClass "add" argTypes)
               #!null
               (Object[] (Integer 1) (Integer 2)))))
(format #t "Result of HelloWorld.add(1,2) is ~s~%" result)
