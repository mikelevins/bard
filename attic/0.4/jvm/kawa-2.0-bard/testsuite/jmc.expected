
Access flags: public
This class: HelloWorld super: java.lang.Object
Interfaces (count: 0):

Fields (count: 0):

Methods (count: 1):

Method name:"add" public static Signature: (int,int)int
Attribute "Code", length:36, max_stack:2, max_locals:3, code_length:6
  0: iload_0
  1: iload_1
  2: iadd
  3: dup
  4: istore_2
  5: ireturn
Attribute "LocalVariableTable", length:12, count: 1
  slot#2: name: result, type: int (pc: 0 length: 6)

Attributes (count: 0):
Result of HelloWorld.add(1,2) is 3
